### PR TITLE
Generic Message Compression

### PR DESCRIPTION
**Pull request by cchengleo**
This series commits implement a prototype for a generic message compression mechanism based on https://github.com/ceph/ceph/pull/3303 and commons. However, the interface to make use of this is still not clear.

### CODE REVIEW

**Comment by mattbenjamin** [comment]
Is this also something it could make sense to enable/disable for a connection or flow, rather than message by message?

**Comment by cchengleo** [comment]
@mattbenjamin I agree with you. I think a per connection compression flag can save bandwidth a lot when communicating with remote OSD for site backup or something else.

**Comment by yuyuyu101** [comment]
I'm not sure, maybe some messages which can't get great compress effect doesn't need to send with compress enabled. I think whether we compress a message isn't decided by connection, and should be considered by message itself. For this case, I prefer to add flag to message footer flag.

**Comment by mattbenjamin** [comment]
That makes sense, and I agree.  I'm trying to think of ways to pull conditions out of the decode path...

**Comment by liewegas** [comment]
> Is this also something it could make sense to enable/disable for a connection or flow, rather than message by message?

I can imagine that some messages are more compressible than others...

**Comment by cchengleo** [comment]
@markhpc Could you please share more information on your question? What aspects do you mean for "take it and pass it around without a decompression and re-compression step..."?

**Comment by cchengleo** [comment]
@yuyuyu101 I have already make use of **footer.flags** to tag compressed messages. We can let upper levels to make the decision and set **footer.flags** earlier.

File: src/include/msgr.h [hunk 1]
```diff
@@ -188,6 +190,7 @@ struct ceph_msg_footer {
 #define CEPH_MSG_FOOTER_COMPLETE  (1<<0)   /* msg wasn't aborted */
 #define CEPH_MSG_FOOTER_NOCRC     (1<<1)   /* no data crc */
 #define CEPH_MSG_FOOTER_SIGNED	  (1<<2)   /* msg was signed */
+#define CEPH_MSG_FOOTER_COMPRESS  (1<<3)   /* msg was compressed */
```
Comment (liewegas) [inline src/include/msgr.h]: I think a better way to approach this is a flag in the header, CEPH_MSG_HEADER_COMPRESSDATA and COMPRESSFRONT and COMPRESSMIDDLE (or whatever the naming scheme is).  And make the first 4 bytes of each of those payloads be the decompressed length.  This way we don't change any of the fundamental msg structures.

File: src/msg/Message.cc [hunk 2]
```diff
@@ -173,7 +173,7 @@ using namespace std;
 
 #define dout_subsys ceph_subsys_ms
 
-void Message::encode(uint64_t features, bool datacrc)
+void Message::encode(uint64_t features, bool datacrc, bool compression)
```
Comment (liewegas) [inline src/msg/Message.cc]: we'll want to rebase this on top of the patch that changes this bool to a flags field.

Comment (liewegas) [inline src/msg/Message.cc]: https://github.com/ceph/ceph/pull/3473

**Comment by liewegas** [comment]
in the final version you'll need to drop teh first patch (or squash it in to the next one) so that we don't add the recovery-specific code and then remove it in the next patch

**Comment by liewegas** [comment]
I think a sensible approach would be to add Message methods like

 virtual should_compress() { return false; }
 virtual should_not_compress() { return false; }

That way we can flag any messages that are good or bad candidates.  Then we can take a policy that only compressings things that are explicitly flagged as good candidates, or compress everything that is not flagged as a bad candidate?

We'll also need to add a feature bit for compatibility, but we can do that last...

**Comment by cchengleo** [comment]
@liewegas  `in the final version you'll need to drop teh first patch (or squash it in to the next one) so that we don't add the recovery-specific code and then remove it in the next patch`

I have removed recovery-specific code originally coding by @XinzeChi in https://github.com/cchengleo/ceph/commit/81d64c5814be1e9ba97187d5c1a6f3a3d56c3fb0

**Comment by cchengleo** [comment]
@liewegas `I think a sensible approach would be to add Message methods like

virtual should_compress() { return false; }
virtual should_not_compress() { return false; }

That way we can flag any messages that are good or bad candidates. Then we can take a policy that only compressings things that are explicitly flagged as good candidates, or compress everything that is not flagged as a bad candidate?

We'll also need to add a feature bit for compatibility, but we can do that last...`

Could you please elaborate this a little bit? Are higher level logics use these API to set flag for compression?

**Comment by liewegas** [comment]
It looks to me like you have the feature bit handled correctly for compatibility.

My main concern is that having a Message::compress() and decompress() method adds another dimension of state to the Message struct, which means that other code may potentially need to cope or understand with a compressed Message (or even just that we need to meaningfully represent a compressed Message in memory).  Maybe we want to go down that road.... I'm not sure.

But I think a simpler approach would be what I describe in an earlier comment:

1. the msg header struct has the original crc and length values
2. the msg header has bits set if the over-the-wire payload is compressed
3. the msg/simple/Pipe.cc read_message() and write_message() methods have an alternate path that compresses/decompresses the payload right then in private buffers in the local scope.
4. the over-the-write compressed buffer (whether it is front, middle, or data) has the compressed crc, length, and then compressed payload.

Perhaps in the future if we want to represent the compressed message as a valid Message we can tackle that.. but if we avoid it entirely for now I think it's safer and simpler!

**Comment by liewegas** [comment]
An open question for me is whether we leave the header flags for compression set on the Message that we received.  This tells others that the data was once compressed on the wire, which might conceivably be interesting, but is someone confusing since it the buffers are no longer compressed.  Similarly, if we are sending a message that has those flags set, does that mean we are requesting or forcing compression?  We'll need to make sure they are cleared before sending over the wire if, for example, the other end doesn't support compression.

**Comment by cchengleo** [comment]
@liewega

**Comment by gregsfortytwo** [comment]
I don't think we should leak the compressed state outside of the messenger, if that's what handles encode/decode.

A more integrated compression system (that, eg, keeps the data compressed until it's passed into the ObjectStore, or even writes it to disk and transparently handles decompressing on update)) might be interesting but a lot more work that I'm not sure is worth it (the reason to do it would be eg memory bandwidth on a system that needs to copy data between threads or something). Outside of a scenario where the external system can actually change its behavior in meaningful ways based on compression state it should be invisible.

**Comment by cchengleo** [comment]
@liewegas 
On **read_message**, **front/middle/data_len** are used to get read front/middle/data correctly over-the-wire. These fields represents the real length of the following payloads. It also applies to front/middle/data_crc. That's why I didn't emit any *original metadata* in the header field.

**Comment by liewegas** [comment]
Yeah, I think what you have now works.  I'm suggesting that keeping those fields populated by original values will avoid the swicheroo with the throttler, keep the compression logic super-localized (to just read_message and write_message), and make things generally simpler..

**Comment by cchengleo** [comment]
@liewegas 
I agree with *keep compression logic super-localized)*. My original intention to put compression logic in **Message** is because I want to handle both AsynConnection and Pipe case.

**Comment by cchengleo** [comment]
@liewegas 
But if we leave compression logic "super-localized" in **write_message**, then we have to redo extra work such as recalculating header/payloads/footer CRC.

**Comment by liewegas** [comment]
I see, that's a good point.

Looking at the code now I see that the write_message caller does this:

<code>
	bufferlist blist = m->get_payload();
	blist.append(m->get_middle());
	blist.append(m->get_data());

        pipe_lock.Unlock();

        ldout(msgr->cct,20) << "writer sending " << m->get_seq() << " " << m << dendl;
	int rc = write_message(header, footer, blist);
</code>

Maybe we replace the blist construction with something like Message::construct_over_the_wire_bl(bufferlist *outbl)?  The read size would still be a bit of a mess, though.

@yuyuyu101 what do you think?

**Comment by cchengleo** [comment]
@liewegas 
Do you mean to compress in **Message::construct_over_the_wire_bl(bufferlist *outbl)**?

**Comment by liewegas** [comment]
Yeah.  Or otherwise make a helper that does the compression that can be used by both messengers but doesn't affect the in-memory representation of Message

**Comment by liewegas** [comment]
(i.e., make it a const method...)

**Comment by cchengleo** [comment]
@liewegas 
I see! If the purpose is to preserve in-memory representation of Message, does that mean we can have **Message::compress(bufferlist &header, bufferlist &middle, bufferlist &data)** in **write_message**?

**Comment by liewegas** [comment]
On Tue, 27 Jan 2015, Cheng Cheng wrote:
> 
> @liewegas
> I see! If the purpose is to preserve in-memory representation of Message, does that mean we can have Message::compress(bufferlist &header, bufferlist &middle, bufferlist &data) in write_message?

Exactly (but with a bufferlist& front in there too).

Unfortunately that doesn't really solve the read side, which is a mess.  
:/

**Comment by cchengleo** [comment]
@liewegas 
If we are going to take that approach, it implies that even the message has been signed, we still perform compression before sending to the wire. On RX, **read_message** should eventually have a message with exactly same in-memory representation as the sender's before compression?

**Comment by liewegas** [comment]
On Tue, 27 Jan 2015, Cheng Cheng wrote:
> @liewegas
> If we are going to take that approach, it implies that even the message has been signed, we still perform compression before sending to the wire. On RX, read_message should eventually have a message with exactly same in-memory
> representation as the sender's before compression?

Yeah, I think so!

**Comment by yuyuyu101** [comment]
I would like to stress more, because "seq" in header is a contention field that AsyncMessenger permit multi thread access send_message. And crc calculation will consume lots of time, I hope we can separate crc calculation especially header crc calculation. AsyncMessenger can calculate header crc very late and lock can be in a very short path instead of locking during crc calculation.

**Comment by markhpc** [comment]
cchengleo:  More or less what Greg said.  At some point we might want to be able to take comrpessed data (flag it?) and transmit it across the wire "as-is" for some other layer to decompress.  I agree with him though that we shouldn't leak anything going on inside the messenger (or other peice) outside ceph.  Probably not relevant to this conversation, except perhaps that it might be nice to disable compression if we can hint that a particular peice of data is already compressed.

**Comment by cchengleo** [comment]
@markhpc 
If we compress/decompress at a specific operation layer, then yes, we would not leak the compressed state outside of the messenger, because that upper layer knows how to interoperate data already. This seems relevant to what https://github.com/ceph/ceph/pull/3303 proposed. Otherwise, we have to carry a flag in the header to indicate the message payloads are compressed and need to be decompressed for further processing.

File: debian/control [hunk 3]
```diff
@@ -10,6 +10,7 @@ Build-Depends: autoconf,
                automake,
                autotools-dev,
 	       libbz2-dev,
+               liblz4-dev,
```
Comment (dachary) [inline debian/control]: this must also go in the ceph.spec.in BuildDepends so that RPM based platforms install it (centos being one of them ;-)

**Comment by dachary** [comment]
<pre>
Error: No Package found for liblz4-devel
</pre>
This is lz4-devel (no lib prefix) on centos.

**Comment by majianpeng** [comment]
I wanted to compress the journal for large write(from my test, (My cluster setup is five osd w/ on ssd as journal. w/o ec & replica size = 1, all data-zero the performance increase 44%. For random-data the performance increate 37%). But at present, there is no compress func/class.
@cchengleo  Can you firstly make the first patch merge into the master. After that, i can don my journal compress/decompress.  Thanks!

**Comment by cchengleo** [comment]
@majianpeng Okay I will put this piece in my TODO list. And make run-make-check pass. From the design aspect, I think the current one is okay to perform necessary functionality.

**Comment by dachary** [comment]
@cchengleo feel free to re-open when ready.

**Comment by yuyuyu101** [comment]
I heard someone could benefit network message compress for recovery and some use case like vdi. So I would like to push this again. @cchengleo Would you have a new impl or reopen it again to make complete?

**Comment by cchengleo** [comment]
I can follow up this PR. Let me take a look for current network impl and make changes to accommodate. I also would like to make it complete.

Sincerely,
Cheng

**Comment by yuyuyu101** [comment]
@cchengleo I hope we can get some progress for this :-) I would like to help to complete this PR if you have any  problem.

**Comment by yuyuyu101** [comment]
Actually we may like to do compress even in messenger level async. Online compress within messenger thread may affect network bandwidth hugely

**Comment by cchengleo** [comment]
@yuyuyu101 Thanks for the help! I am recently get busy with other stuffs. Please go ahead and implement what you proposed. I will try to help also.

I am not quite get why online compressing affect network bandwidth hugely. Can you further elaborate?

**Pull request closed without merging**